File: algorithm/study/1331.py
D = {'A': 0, 'B': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5,
     '1': 0, '2': 1, '3': 2, '4': 3, '5': 4, '6': 5}


dr = [[1, -2], [2, -1], [2, 1], [1, 2], [-1, 2], [-2, 1], [-2, -1], [-1, -2]]

vl = [[False for _ in range(6)] for _ in range(6)]
nl = [input() for _ in range(36)]
nl.append(nl[0])
cx, cy = D[nl[0][0]], D[nl[0][1]]
cnt = 0
for i in range(1, 37):
    for dx, dy in dr:
        tx, ty = cx + dx, cy + dy
        if tx == D[nl[i][0]] and ty == D[nl[i][1]] and not vl[ty][tx]:
            cx, cy = tx, ty
            cnt += 1
            vl[ty][tx] = True
            break
    if cnt != i:
        break

if cnt != 36:
    print('Invalid')
else:
    print('Valid')

File: algorithm/백준/1331.cpp
#include <stdio.h>
char ml[6][6] = { 0, };
char position[3];
char tx, ty;
char res = 1;
bool chk(char cx, char cy, char nx, char ny) {
	if (cx - nx == 2 || nx - cx == 2) {
		if (cy - ny == 1 || ny - cy == 1) return 1;
		return 0;
	}
	else if (cx - nx == 1 || nx - cx == 1) {
		if (cy - ny == 2 || ny - cy == 2) return 1;
		return 0;
	}
	return 0;
}

int main() {
	scanf("%s", &position);
	tx = position[1];
	ty = position[0];
	int cnt = 35;
	char initPosition[2];
	initPosition[0] = position[0];
	initPosition[1] = position[1];
	while (cnt--) {
		scanf("%s", &position);
		if (chk(tx, ty, position[1], position[0])) {
			if (ml[position[0] - 'A'][position[1] - '1']) {
				res = 0;
				break;
			}
			ml[position[0] - 'A'][position[1] - '1'] = 1;
		}
		else {
			res = 0;
			break;
		}
		tx = position[1];
		ty = position[0];
	}
	if (res && !ml[initPosition[1]][initPosition[0]]){
		if (chk(tx, ty, initPosition[1], initPosition[0])) {
			printf("Valid");
			return 0;
		}
	}
	printf("Invalid");
	return 0;
}
